### PR TITLE
[ci] Update CUDA toolkit to v12.1.1

### LINKED ISSUE #13830: move CUDART default version to use 12.1 from 11.6.2
### Request description

Currently we download and use CUDA 11.6.2.  Filing a tracker bug to move us to 12.1.1

We need 12.1.1 (https://developer.download.nvidia.com/compute/cuda/redist/redistrib_12.1.1.json) for Aarch64 CPU + CUDA.  11.6.2 doesn't expose aarch64 builds for CUDA. 

### What component(s) does this issue relate to?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by antiagainst**
CUDA 12 enables users to target architecture-specific features in Hopper and Ada Lovelace architectures. It also enhances Graphs API.

Fixes https://github.com/openxla/iree/issues/13830

### CODE REVIEW

**Comment by stellaraccident** [comment]
Ping when this is close to being worked out. Looks like still issues.

File: build_tools/docker/dockerfiles/nvidia.Dockerfile
Comment (pzread) [inline build_tools/docker/dockerfiles/nvidia.Dockerfile:1]: Line 55 also installs CUDA toolkit. We might need to also update it?

File: build_tools/docker/dockerfiles/nvidia-bleeding-edge.Dockerfile
Comment (pzread) [inline build_tools/docker/dockerfiles/nvidia-bleeding-edge.Dockerfile:1]: Same in this file on line 64, we also install CUDA toolkit there.

**Comment by antiagainst** [comment]
Okay, after taking a closer look this is more complicated than just bump a number. :) From https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-toolkit-major-component-versions, CUDA 12.1.x requires the driver version to be >=525.60.13, while we are using 460 for the normal dockers, and 515 for bleeding edge dockers. So those all needs to be updated, which are intertwined with the Vulkan drivers. We might also need to update the host machine's drivers too..

**Comment by stellaraccident** [comment]
> Okay, after taking a closer look this is more complicated than just bump a number. :) From https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-toolkit-major-component-versions, CUDA 12.1.x requires the driver version to be >=525.60.13, while we are using 460 for the normal dockers, and 515 for bleeding edge dockers. So those all needs to be updated, which are intertwined with the Vulkan drivers. We might also need to update the host machine's drivers too..

Yeah, figured. We can solicit further help on some of these upgrades...

**Comment by powderluv** [comment]
We are running with Ubuntu 22.04 + CUDA 12.1 on the SHARK setups. A few gotchas

GCP latest kernel for Ubuntu 22.04 was bumped from 5.15 to 5.19 which was compiled with GCC 12.x . So you will need gcc-12 installed before your install cuda-12.1x (with the 530 drivers). We also had to build / install latest nccl.  The vulkan setup was automatic. 

```
:~/github/iree$ uname -a
Linux ubuntu 5.19.0-1022-gcp #24~22.04.1-Ubuntu SMP Sun Apr 23 09:51:08 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
:~/github/iree$ nvidia-smi | grep Version
| NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1     |
```

Happy to help create the VM image or help in anyway. 

Also  should we consider _not_ pinning docker images to a SHA and instead rely on the CI / Release build printing out the SHA of a particular run / build and we just track latest always. We can go back to a SHA a build used at any time. This avoids a black magic 26 docker image update because we had to bump a minor revision of CUDA in the future.  e.g: PyTorch uses an 18.04 "latest" image https://github.com/pytorch/builder/blob/60c79a85fb3a1dbe327a735f1f03e7f110ca0427/libtorch/build_docker.sh#LL24C53-L24C53  and it is not pinned to a SHA.  I understand the need to track the exact image / SHA but could we try to relax that for simplicity and see how far it goes.

**Comment by GMNGeoffrey** [comment]
I'm not sure which sha pinning you're talking about. If we were using some upstream containers someone else was publishing then I think this would maybe "just work" at the cost of having failures marked on a commit that have nothing to do with that commit. But I don't see how it would work well for own docker containers.

The change to use digests everywhere was quite deliberate and based on getting somewhat reproducible builds of the docker containers themselves, not just the CI builds using docker. Previously it was quite difficult to reliably update the docker images, have such an update reviewed, test that update on presubmit, roll back a bad update, etc. Some discussion here: https://github.com/openxla/iree/issues/3117

Also isn't this is a Cuda major revision? I don't think we want that to be something that happens automagically. I agree with your general point that having to rebuild all the docker images is annoying (especially since we haven't got it configured right to cache the intermediate layers in the registry), but if we weren't also updating the base OS, then updating the Cuda version should require rebuilding only the three docker images that install Cuda. To the extent that we actually need three separate images there, that's as minimal as it could get.

I'm happy for a rework of our Docker usage (or really I'd love to use some other container system because Dockerfile is awful), but I don't think just taking what we have now, but make it less hermetic, is the way to go.

**Comment by antiagainst** [comment]
> So you will need gcc-12 installed before your install cuda-12.1x (with the 530 drivers). We also had to build / install latest nccl.

@powderluv Bots are all happy. I didn't see any errors. So I'm not sure whether the above still applies?

**Comment by antiagainst** [comment]
@stellaraccident @GMNGeoffrey @pzread This is ready for review now. PTAL. Thanks!

**Review by GMNGeoffrey — APPROVED** [review]
Some people (like me) don't have an Nvidia GPU on their dev machine, but still build the cuda target and driver. I think we use a forked parse_redist.py somewhere to fetch it during the build. Can you verify that this still works?

**Comment by antiagainst** [comment]
> Some people (like me) don't have an Nvidia GPU on their dev machine, but still build the cuda target and driver. I think we use a forked parse_redist.py somewhere to fetch it during the build. Can you verify that this still works?

Ha. Yes, turns out we need to install another component with `parse_redist.py` for `#include <nv/target>`. Thanks!

**Comment by stellaraccident** [comment]
> Some people (like me) don't have an Nvidia GPU on their dev machine, but still build the cuda target and driver. I think we use a forked parse_redist.py somewhere to fetch it during the build. Can you verify that this still works?

Some of us have it on our dev machine and still use this route :)

**Review by stellaraccident — APPROVED** [review]

**Pull request merged**